src/game-engine/application-states/Base.ts

```diff
@@ -1,20 +1,34 @@
-import { Channel } from 'fmg-core';
-
-export default class BaseState {
-  balances: number[];
-  channel: Channel;
-  stake: number;
-  type: string;
-  playerIndex: number; // overwritten by subclass
+import { Player } from '.';
+import { Position } from '../positions';
+
+export default class BaseState<T extends Position> {
+  readonly position: T;
```

Comment:
What is the `T` for? can't all of these just be `Position`?

Reply:
I had to do this to allow `ApplicationState` subclasses to be specialised to certain position subclasses, e.g:
```
export class WaitForResting extends BasePlayerA<Reveal> {
  readonly type = PlayerAStateType.WAIT_FOR_RESTING;
  readonly isReadyToSend = false;

  get stake() { return this.position.stake; }
  get aPlay() { return this.position.aPlay; }
  get bPlay() { return this.position.bPlay; }
  get salt() { return this.position.salt; }
  get result() { return calculateResult(this.aPlay, this.bPlay); }
}
```
If you can't specify that the `position` in the `WaitForResting` state is actually a `Reveal` position, you can't call e.g. `this.position.aPlay`, because not all `Positions` have an `aPlay` property.

Reply:
Ahhh, great - makes sense. Thanks for the explanation.